hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndex.scala

```diff
@@ -224,6 +226,126 @@ class TestPartitionStatsIndex extends PartitionStatsIndexTestBase {
+      saveMode = SaveMode.Append)
+    doWriteAndValidateDataAndPartitionStats(
+      hudiOpts,
+      operation = DataSourceWriteOptions.UPSERT_OPERATION_OPT_VAL,
+      saveMode = SaveMode.Append)
+    // Clean Operation
+    val lastCleanInstant = getLatestMetaClient(false).getActiveTimeline.getCleanerTimeline.lastInstant()
+    assertTrue(lastCleanInstant.isPresent)
+    // do another upsert and validate the partition stats including file pruning
+    doWriteAndValidateDataAndPartitionStats(
+      hudiOpts,
+      operation = DataSourceWriteOptions.UPSERT_OPERATION_OPT_VAL,
+      saveMode = SaveMode.Append)
+
+    validateDataAndPartitionStats()
+    createTempTable(hudiOpts)
+    verifyQueryPredicate(hudiOpts)
+  }
+
+  /**
+   * Test case to do updates and then validate partition stats with clustering.
+   */
+  @ParameterizedTest
+  @EnumSource(classOf[HoodieTableType])
+  def testPartitionStatsWithClustering(tableType: HoodieTableType): Unit = {
+    val hudiOpts = commonOpts ++ Map(
+      DataSourceWriteOptions.TABLE_TYPE.key -> tableType.name(),
+      HoodieClusteringConfig.INLINE_CLUSTERING.key() -> "true",
+      HoodieClusteringConfig.INLINE_CLUSTERING_MAX_COMMITS.key() -> "2",
+      KeyGeneratorOptions.URL_ENCODE_PARTITIONING.key() -> "true")
+
+    doWriteAndValidateDataAndPartitionStats(
+      hudiOpts,
+      operation = DataSourceWriteOptions.INSERT_OPERATION_OPT_VAL,
+      saveMode = SaveMode.Overwrite)
+    doWriteAndValidateDataAndPartitionStats(
+      hudiOpts,
+      operation = DataSourceWriteOptions.UPSERT_OPERATION_OPT_VAL,
+      saveMode = SaveMode.Append)
+    // validate clustering instant
+    val lastClusteringInstant = getLatestClusteringInstant
+    assertTrue(getLatestClusteringInstant.isPresent)
+    // do two more rounds of upsert to trigger another clustering
+    doWriteAndValidateDataAndPartitionStats(hudiOpts,
+      operation = DataSourceWriteOptions.UPSERT_OPERATION_OPT_VAL,
+      saveMode = SaveMode.Append)
+    doWriteAndValidateDataAndPartitionStats(hudiOpts,
+      operation = DataSourceWriteOptions.UPSERT_OPERATION_OPT_VAL,
+      saveMode = SaveMode.Append)
+    assertTrue(getLatestClusteringInstant.get().getTimestamp.compareTo(lastClusteringInstant.get().getTimestamp) > 0)
+    assertEquals(getLatestClusteringInstant, metaClient.getActiveTimeline.lastInstant())
+    // We are validating rollback of a DT clustering instant here
+    rollbackLastInstant(hudiOpts)
+
+    validateDataAndPartitionStats()
+    createTempTable(hudiOpts)
+    verifyQueryPredicate(hudiOpts)
+  }
+
+  /**
+   * Test case to do updates and then validate partition stats with MDT compaction.
+   * Any one table type is enough to test this as we are validating the metadata table.
+   */
+  @Test
+  def testPartitionStatsWithMDTCompaction(): Unit = {
+    val hudiOpts = commonOpts ++ Map(
+      HoodieMetadataConfig.COMPACT_NUM_DELTA_COMMITS.key() -> "1"
```

Comment:
Something bigger than 1 is helpful so the log merging is also exercised.

Reply:
makes sense.. set to 2.